host_core/native/hostcore_wasmcloud_native/src/oci.rs

```diff
@@ -50,32 +53,57 @@ pub(crate) async fn fetch_oci_path(
             "Fetching images tagged 'latest' is currently prohibited in this host. This option can be overridden with WASMCLOUD_OCI_ALLOW_LATEST".into());
     }
     let cf = cached_file(img).await?;
-    if tokio::fs::metadata(&cf).await.is_err() {
-        let auth = determine_auth(img, creds_override);
-        let img = oci_distribution::Reference::from_str(img)?;
 
-        let protocol =
-            oci_distribution::client::ClientProtocol::HttpsExcept(allowed_insecure.to_vec());
-        let config = oci_distribution::client::ClientConfig {
-            protocol,
-            ..Default::default()
-        };
-        let mut c = oci_distribution::Client::new(config);
-        let imgdata = pull(&mut c, &img, &auth).await;
+    let auth = determine_auth(img, creds_override);
+    let img = oci_distribution::Reference::from_str(img)?;
+
+    let protocol =
+        oci_distribution::client::ClientProtocol::HttpsExcept(allowed_insecure.to_vec());
+    let config = oci_distribution::client::ClientConfig {
+        protocol,
+        ..Default::default()
+    };
+    let mut c = oci_distribution::Client::new(config);
 
+    if tokio::fs::metadata(&cf).await.is_err() {     
+        let imgdata = pull(&mut c, &img, &auth).await;
         match imgdata {
             Ok(imgdata) => {
-                let mut f = tokio::fs::File::create(&cf).await?;
-                let content = imgdata
-                    .layers
-                    .into_iter()
-                    .flat_map(|l| l.data)
-                    .collect::<Vec<_>>();
-                f.write_all(&content).await?;
-                f.flush().await?;
+                cache_oci_image(&imgdata, &cf).await;
             }
             Err(e) => return Err(format!("Failed to fetch OCI bytes: {}", e).into()),
         }
+    }else{
+        let manifest = c.pull_manifest(&img, &auth).await;
+        match manifest{
+            Ok(manifest)=>{
+                let (oci_manifest, _) = manifest;
+                if let OciManifest::Image(image_manifest) = oci_manifest {
+                    
+                    let layers = image_manifest.layers;
+                    // The digest will be string like sha256:asdf, and the first 7 characters are removed
+                    let oci_sha = &layers[0].digest[7..];
+                    let oci_sha_bytes: Vec<u8> = hex::decode(oci_sha).expect("Invalid Hex String");
```

Comment:
Oh! One more helpful tip here, you don't actually need to decode as hex if you don't want to. When you hash the file below, you can use `format!("{:x}", hash_result)` and compare the strings. See the example here: https://github.com/deislabs/bindle/blob/f8d20c506e3abab5a5e0fa90e0940d6db84a0908/src/id.rs#L78.

No need to change anything, was just pointing it out for future sanity